src/rgw/rgw_op.h

```diff
@@ -1704,9 +1709,10 @@ static inline void rgw_get_request_metadata(CephContext *cct,
 } /* rgw_get_request_metadata */
 
 static inline void encode_delete_at_attr(ceph::real_time delete_at,
-					map<string, bufferlist>& attrs)
+					map<string, bufferlist>& attrs, bool set_delete_at = false)
 {
-  if (real_clock::is_zero(delete_at)) {
```

Comment:
Yeah. @rzarzynski  : In this change, if we want to set the object expiration time. The var `set_delete_at` will be true and the `delete_at` updated relatively.

In `int RGWObjectExpirer::garbage_single_object(objexp_hint_entry& hint)` it will call `delete_obj()`
In `int RGWRados::Object::Delete::delete_obj()`, you will find these code:
```cpp
if (!real_clock::is_zero(params.expiration_time)) {
    bufferlist bl;
    real_time delete_at;

    if (state->get_attr(RGW_ATTR_DELETE_AT, bl)) {
      try {
        bufferlist::iterator iter = bl.begin();
        ::decode(delete_at, iter);
      } catch (buffer::error& err) {
        ldout(store->ctx(), 0) << "ERROR: couldn't decode RGW_ATTR_DELETE_AT" << dendl;
	return -EIO;
      }

      if (params.expiration_time != delete_at) {
        return -ERR_PRECONDITION_FAILED;
      }
    }
  }
```

In ` if (params.expiration_time != delete_at) {` :
if the expiration time of object is updated, the `delete_at` is the current expiration time in bufferlist and the params,expiration_time is the non-current expiration time in the list of `cls_timeindex_entry`. 

So, that means if the two expriation time is not consistent, the object will not be deleted. 
If we intend to cancel the object expiration time, the `delete_at` will be `0` and the object will never be deleted because of `if (!real_clock::is_zero(params.expiration_time)) {` in upper code.